crates/credentials/src/presentation_definition_v2.rs

```diff
@@ -0,0 +1,230 @@
+    use super::*;
+    use serde_json::json;
+    use std::fs;
+    use std::path::Path;
+
+    #[test]
+    fn can_serialize() {
+        let pd = PresentationDefinitionV2 {
+            id: "tests-pd-id".to_string(),
+            name: "simple PD".to_string().into(),
+            purpose: "pd for testing".to_string().into(),
+            input_descriptors: vec![InputDescriptorV2 {
+                id: "whatever".to_string(),
+                purpose: "purpose".to_string().into(),
+                constraints: ConstraintsV2 {
+                    fields: vec![FieldV2 {
+                        id: "field-id".to_string().into(),
+                        path: vec!["$.issuer".to_string()],
+                        purpose: "purpose".to_string().into(),
+                        filter: json!({"type": "string", "const": "123"}).into(),
+                        ..Default::default()
+                    }]
+                    .into(),
+                    limit_disclosure: Some(ConformantConsumerDisclosure::Required),
+                },
+                ..Default::default()
+            }],
+            ..Default::default()
+        };
+
+        let serialized_pd = serde_json::to_string(&pd).unwrap();
+
+        assert!(serialized_pd.contains("input_descriptors"));
+        assert!(serialized_pd.contains("123"));
+    }
+
+    #[test]
+    fn can_deserialize() {
+        let expected_id = "ec11a434-fe24-479b-aae0-511428b37e4f";
+
+        let expected_format = Format {
+            jwt_vc: Some(JwtObject {
+                alg: vec!["ES256K".to_string(), "EdDSA".to_string()],
+            }),
+            ..Default::default()
+        };
+
+        let expected_input_descriptors = vec![InputDescriptorV2 {
+            id: "7b928839-f0b1-4237-893d-b27124b57952".to_string(),
+            constraints: ConstraintsV2 {
+                fields: Some(vec![
+                    FieldV2 {
+                        path: vec!["$.iss".to_string(), "$.vc.issuer".to_string()],
+                        filter: Some(json!({"type": "string", "pattern": "^did:[^:]+:.+"})),
+                        ..Default::default()
+                    },
+                    FieldV2 {
+                        path: vec!["$.vc.type[*]".to_string(), "$.type[*]".to_string()],
+                        filter: Some(json!({"type": "string", "const": "SanctionsCredential"})),
+                        ..Default::default()
+                    },
+                ]),
+                ..Default::default()
+            },
+            ..Default::default()
+        }];
+
+        let pd_string = load_json("tests/resources/pd_sanctions.json");
+        let deserialized_pd: PresentationDefinitionV2 = serde_json::from_str(&pd_string).unwrap();
+
+        assert_eq!(deserialized_pd.id, expected_id);
+        assert_eq!(deserialized_pd.format, Some(expected_format));
+        assert_eq!(
+            deserialized_pd.input_descriptors,
+            expected_input_descriptors
+        );
```

Comment:
I think this is the most powerful test. In kt, it checks that JSON is equal (see [serialize / deserialize idempotency](https://github.com/TBD54566975/web5-kt/blob/main/credentials/src/test/kotlin/web5/sdk/credentials/PresentationDefinitionTest.kt#L56)). Would it make sense to compare the the actual JSON outputs from the pre-serialization JSON vs the serialized and then deserialized JSON?

Reply:
I couldn't find a canonicalizer written in Rust that behaves the same way as the Kotlin test. Whitespace differences caused issues when comparing the json strings.

That said, I did vastly improve the deserialize test to ensure that the JSON is parsed into the expected format here, whereas Kotlin is just testing that it doesn't throw when deserializing. It _should_ be testing the same thing.

Reply:
Curious if you took a look into https://docs.rs/serde_jcs/latest/serde_jcs/ ?

Reply:
Yes. Multiple times. Works well, but I couldn't get it to match white space between json of just the string vs deserialized jcs. I don't know why yet. In all honesty, I don't think it's worth fretting over yet. I'll make a task in the backlog to figure it out.

Reply:
Sounds good! I'd love to take a look at that as a first stab at rust!